### ISSUE TITLE
`tf.experimental.numpy.var` Causes `SIGABRT` on Empty Tensors with Large Dimensions

### ISSUE THREAD

**Issue by jiren-the-gray** [tensorflow/tensorflow#108921]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

binary

### TensorFlow version

2.20.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 24.04

### Python version

3.12

### Current behavior?

A `CHECK` failure (`SIGABRT`) occurs in `tensorflow/core/framework/tensor_shape.cc` when performing a reduction operation (specifically `tf.experimental.numpy.var`) on a tensor that contains both a 0 dimension and a dimension close to INT64_MAX.

Although the tensor is logically empty (0 elements), the internal shape inference logic appears to calculate the product of dimensions before accounting for the zero dimension. This calculation overflows int64, resulting in a negative value for the number of elements, which triggers the assertion `Check failed: 0 <= new_num_elements`.

Here is a [gist](https://colab.research.google.com/gist/jiren-the-gray/93c6cf7b85e6fcca9ecf00aaaf188fd6/tf-experimental-numpy-var_crash.ipynb)

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
import numpy as np

print("TensorFlow version:", tf.__version__, flush=True)

# 1. Create a tensor with a 0 dimension (empty) and a huge dimension.
# Shape logic: 2 * 2 * (2^63 - 2) overflows signed int64.
# The '0' dimension prevents actual memory allocation but preserves the shape metadata.
huge_dim = 9223372036854775806  # (2^63 - 2)
shape = (2, 2, 0, huge_dim)

tensor = tf.zeros(shape=shape, dtype=tf.int64)

# 2. Run variance reduction.
# This triggers the internal reshape/reduction logic that hits the CHECK failure.
# ddof=1 is required to match the crash path.
tf.experimental.numpy.var(tensor, axis=2, ddof=1)

print("Print this if no crash occurs.", flush=True)
```

### Relevant log output

```shell
TensorFlow version: 2.20.0
WARNING: All log messages before absl::InitializeLog() is called are written to STDERR
I0000 00:00:1769492950.420590  421930 gpu_device.cc:2020] Created device /job:localhost/replica:0/task:0/device:GPU:0 with 21934 MB memory:  -> device: 0, name: NVIDIA GeForce RTX 4090, pci bus id: 0000:01:00.0, compute capability: 8.9
2026-01-27 00:49:10.592813: F tensorflow/core/framework/tensor_shape.cc:421] Check failed: 0 <= new_num_elements (0 vs. -1)
Aborted (core dumped)
```

**Comment by jiren-the-gray** [comment]
The same bug is observed in these additional APIs: `tf.experimental.numpy.all, tf.experimental.numpy.any, tf.experimental.numpy.amax, tf.experimental.numpy.amin, tf.experimental.numpy.count_nonzero, tf.experimental.numpy.max, tf.experimental.numpy.nanmean, tf.experimental.numpy.ptp, tf.experimental.numpy.nansum, tf.experimental.numpy.reshape, tf.experimental.numpy.promote_types, tf.experimental.numpy.var, tf.experimental.numpy.std`

**Comment by Kayyuri** [comment]
Hi @jiren-the-gray, Thank you for raising the issue.
I was able to reproduce the same error with `TensorFlow 2.20` and `tf-nightly 2.21.0-dev20260202` . I am attaching [gist](https://colab.sandbox.google.com/gist/Kayyuri/9c562ea9d58329d8955f0f6acb7f3100/108921.ipynb) file for the reference. Thanks!

**Comment by Kayyuri** [comment]
Hi @jiren-the-gray,
Thank you so much for filing this issue! We truly appreciate your support. At the moment, this request is not prioritized and mark it for community contributions. If you would like to contribute on a fix, please feel free to fork the repository and submit a pull request in accordance with the TensorFlow [CONTRIBUTION GUIDELINES](https://www.tensorflow.org/community/contribute/code) and link this issue in your PR to ensure the maintainers can easily review and collaborate.
To avoid duplicate work, please leave a comment here if you decide to take it on!

Thank you again for your interest and we look forward to seeing your contributions!

**Comment by vishwakt** [comment]
I've opened PR #126019 with a fix for this.

The crash is not specific to `tf.experimental.numpy.var`. Plain `tf.reduce_max(tf.zeros((2, 2, 0, 2**63 - 2), tf.int64), axis=2)` aborts with the identical stack, which lines up with the list of affected APIs in your follow-up, and it is the same underlying bug as #108891.

A TensorShape is built one dimension at a time and only rejected when a prefix of it overflows, so `(2, 2, 0, 2**63 - 2)` is accepted: the zero dimension drives the running element count to zero before the large dimension is applied. `ReductionHelper::Simplify` in tensorflow/core/kernels/reduction_ops_common.cc then computes the output shape by dropping the reduced dimension, which leaves `(2, 2, 2**63 - 2)`, and turning that back into a TensorShape trips the non-negative element count check and aborts.

The PR computes the output element count with MultiplyWithoutOverflow and returns InvalidArgument instead. Simplify already returns a Status and is called under OP_REQUIRES_OK, so the abort becomes an ordinary op error.

**Comment by Kayyuri** [comment]
Hi @jiren-the-gray, A fix for this issue has been proposed in this PR(#126019). Once it is merged, this issue will be resolved.

Hi @vishwakt ,
Thank you for raising the PR to address this issue. We truly appreciate your initiative and the effort in contributing the fix.
